src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -1138,365 +1708,127 @@ pub fn is_useful<'p, 'a, 'tcx>(
-fn constructor_arity(cx: &MatchCheckCtxt<'a, 'tcx>, ctor: &Constructor<'tcx>, ty: Ty<'tcx>) -> u64 {
-    debug!("constructor_arity({:#?}, {:?})", ctor, ty);
-    match ty.kind {
-        ty::Tuple(ref fs) => fs.len() as u64,
-        ty::Slice(..) | ty::Array(..) => match *ctor {
-            Slice(length) => length,
-            ConstantValue(_) => 0,
-            _ => bug!("bad slice pattern {:?} {:?}", ctor, ty)
-        }
-        ty::Ref(..) => 1,
-        ty::Adt(adt, _) => {
-            adt.variants[ctor.variant_index_for_adt(cx, adt)].fields.len() as u64
-        }
-        _ => 0
-    }
-}
-
-/// This computes the types of the sub patterns that a constructor should be
-/// expanded to.
-///
-/// For instance, a tuple pattern (43u32, 'a') has sub pattern types [u32, char].
-fn constructor_sub_pattern_tys<'a, 'tcx>(
-    cx: &MatchCheckCtxt<'a, 'tcx>,
-    ctor: &Constructor<'tcx>,
-    ty: Ty<'tcx>,
-) -> Vec<Ty<'tcx>> {
-    debug!("constructor_sub_pattern_tys({:#?}, {:?})", ctor, ty);
-    match ty.kind {
-        ty::Tuple(ref fs) => fs.into_iter().map(|t| t.expect_ty()).collect(),
-        ty::Slice(ty) | ty::Array(ty, _) => match *ctor {
-            Slice(length) => (0..length).map(|_| ty).collect(),
-            ConstantValue(_) => vec![],
-            _ => bug!("bad slice pattern {:?} {:?}", ctor, ty)
-        }
-        ty::Ref(_, rty, _) => vec![rty],
-        ty::Adt(adt, substs) => {
-            if adt.is_box() {
-                // Use T as the sub pattern type of Box<T>.
-                vec![substs.type_at(0)]
-            } else {
-                adt.variants[ctor.variant_index_for_adt(cx, adt)].fields.iter().map(|field| {
-                    let is_visible = adt.is_enum()
-                        || field.vis.is_accessible_from(cx.module, cx.tcx);
-                    if is_visible {
-                        let ty = field.ty(cx.tcx, substs);
-                        match ty.kind {
-                            // If the field type returned is an array of an unknown
-                            // size return an TyErr.
-                            ty::Array(_, len)
-                                if len.try_eval_usize(cx.tcx, cx.param_env).is_none() =>
-                                cx.tcx.types.err,
-                            _ => ty,
-                        }
-                    } else {
-                        // Treat all non-visible fields as TyErr. They
-                        // can't appear in any other pattern from
-                        // this match (because they are private),
-                        // so their type does not matter - but
-                        // we don't want to know they are
-                        // uninhabited.
-                        cx.tcx.types.err
-                    }
-                }).collect()
-            }
+        PatKind::Or { ref pats } => {
+            pats.iter().flat_map(|pat| pat_constructors(tcx, param_env, pat, ty)).collect()
```

Comment:
Since you've familiarized yourself in-depth with this code, would you like to tackle exhaustiveness for or-patterns as well? (Hope you don't mind @dlrobertson -- then you can focus more on MIR)

Reply:
Yeah, don't mind at all. Having more eyes on the code is always better :smile:

Reply:
I would like to yes !